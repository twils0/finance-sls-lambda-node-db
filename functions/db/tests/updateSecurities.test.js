const { updateSecurities } = require('../updateSecurities');

global.pool = {
  query: jest.fn(),
};

const userId = 'testUserID';
const securityId1 = 'testSecurityID1';
const securityId2 = 'testSecurityID2';
const securityId3 = 'testSecurityID3';
const body = {
  current: securityId2,
  list: [securityId1, securityId2, securityId3],
};
const indexArray = body.list.map((id, index) => index);
const emptyResponse = {
  rows: [],
};
const noMatchUserResponse = {
  rows: [
    {
      security_current: securityId1,
    },
  ],
};
const matchUserResponse = {
  rows: [
    {
      security_current: securityId2,
    },
  ],
};
const noMatchJoinResponse = {
  rows: [
    {
      security_id: securityId1,
    },
    {
      security_id: securityId2,
    },
  ],
};
const matchJoinResponse = {
  rows: [
    {
      security_id: securityId1,
    },
    {
      security_id: securityId2,
    },
    {
      security_id: securityId3,
    },
  ],
};

describe('functions', () => {
  describe('db', () => {
    describe('updateSecurities', () => {
      afterEach(() => {
        global.pool.query.mockReset();
      });

      it('fails and returns unexpected error, when global.pool.query throws error first query', async () => {
        const error = {
          code: 'testCode',
          message: 'testMessage',
        };
        global.pool.query.mockImplementation((dbString, valuesArray) => {
          expect(dbString).toMatchSnapshot();
          expect(valuesArray).toEqual([userId]);
          return Promise.reject(error);
        });

        try {
          await updateSecurities(userId, body);
        } catch (errorCatch) {
          expect(errorCatch).toEqual(error);
        }
      });

      it('fails and returns unexpected error, when global.pool.query throws error second query', async () => {
        const error = {
          code: 'testCode',
          message: 'testMessage',
        };
        global.pool.query.mockImplementationOnce((dbString, valuesArray) => {
          expect(dbString).toMatchSnapshot();
          expect(valuesArray).toEqual([userId]);
          return Promise.resolve(noMatchUserResponse);
        });
        global.pool.query.mockImplementation((dbString, valuesArray) => {
          expect(dbString).toMatchSnapshot();
          expect(valuesArray).toEqual([userId]);
          return Promise.reject(error);
        });

        try {
          await updateSecurities(userId, body);
        } catch (errorCatch) {
          expect(errorCatch).toEqual(error);
        }
      });

      it('fails and returns unexpected error, when global.pool.query throws error third query', async () => {
        const error = {
          code: 'testCode',
          message: 'testMessage',
        };
        global.pool.query.mockImplementationOnce((dbString, valuesArray) => {
          expect(dbString).toMatchSnapshot();
          expect(valuesArray).toEqual([userId]);
          return Promise.resolve(noMatchUserResponse);
        });
        global.pool.query.mockImplementationOnce((dbString, valuesArray) => {
          expect(dbString).toMatchSnapshot();
          expect(valuesArray).toEqual([userId]);
          return Promise.resolve(noMatchJoinResponse);
        });
        global.pool.query.mockImplementation((dbString, valuesArray) => {
          expect(dbString).toMatchSnapshot();
          expect(valuesArray).toEqual([userId, body.current]);
          return Promise.reject(error);
        });

        try {
          await updateSecurities(userId, body);
        } catch (errorCatch) {
          expect(errorCatch).toEqual(error);
        }
      });

      it('fails and returns unexpected error, when global.pool.query throws error fourth query', async () => {
        const error = {
          code: 'testCode',
          message: 'testMessage',
        };
        global.pool.query.mockImplementationOnce((dbString, valuesArray) => {
          expect(dbString).toMatchSnapshot();
          expect(valuesArray).toEqual([userId]);
          return Promise.resolve(noMatchUserResponse);
        });
        global.pool.query.mockImplementationOnce((dbString, valuesArray) => {
          expect(dbString).toMatchSnapshot();
          expect(valuesArray).toEqual([userId]);
          return Promise.resolve(noMatchJoinResponse);
        });
        global.pool.query.mockImplementationOnce((dbString, valuesArray) => {
          expect(dbString).toMatchSnapshot();
          expect(valuesArray).toEqual([userId, body.current]);
          return Promise.resolve(null);
        });
        global.pool.query.mockImplementation((dbString, valuesArray) => {
          expect(dbString).toMatchSnapshot();
          expect(valuesArray).toEqual([userId]);
          return Promise.reject(error);
        });

        try {
          await updateSecurities(userId, body);
        } catch (errorCatch) {
          expect(errorCatch).toEqual(error);
        }
      });

      it('fails and returns unexpected error, when global.pool.query throws error fifth query', async () => {
        const error = {
          code: 'testCode',
          message: 'testMessage',
        };
        global.pool.query.mockImplementationOnce((dbString, valuesArray) => {
          expect(dbString).toMatchSnapshot();
          expect(valuesArray).toEqual([userId]);
          return Promise.resolve(noMatchUserResponse);
        });
        global.pool.query.mockImplementationOnce((dbString, valuesArray) => {
          expect(dbString).toMatchSnapshot();
          expect(valuesArray).toEqual([userId]);
          return Promise.resolve(noMatchJoinResponse);
        });
        global.pool.query.mockImplementationOnce((dbString, valuesArray) => {
          expect(dbString).toMatchSnapshot();
          expect(valuesArray).toEqual([userId, body.current]);
          return Promise.resolve(null);
        });
        global.pool.query.mockImplementationOnce((dbString, valuesArray) => {
          expect(dbString).toMatchSnapshot();
          expect(valuesArray).toEqual([userId]);
          return Promise.resolve(null);
        });
        global.pool.query.mockImplementation((dbString, valuesArray) => {
          expect(dbString).toMatchSnapshot();
          expect(valuesArray).toEqual([userId, body.list, indexArray]);
          return Promise.reject(error);
        });

        try {
          await updateSecurities(userId, body);
        } catch (errorCatch) {
          expect(errorCatch).toEqual(error);
        }
      });

      it('updates current security, when user table response is empty', async () => {
        global.pool.query.mockImplementationOnce((dbString, valuesArray) => {
          expect(dbString).toMatchSnapshot();
          expect(valuesArray).toEqual([userId]);
          return Promise.resolve(emptyResponse);
        });
        global.pool.query.mockImplementationOnce((dbString, valuesArray) => {
          expect(dbString).toMatchSnapshot();
          expect(valuesArray).toEqual([userId]);
          return Promise.resolve(matchJoinResponse);
        });
        global.pool.query.mockImplementation((dbString, valuesArray) => {
          expect(dbString).toMatchSnapshot();
          expect(valuesArray).toEqual([userId, body.current]);
          return Promise.resolve(null);
        });

        const result = await updateSecurities(userId, body);

        expect(result).toEqual(null);
        expect(global.pool.query).toHaveBeenCalledTimes(3);
      });

      it('updates current security, when securityCurrent does not match', async () => {
        global.pool.query.mockImplementationOnce((dbString, valuesArray) => {
          expect(dbString).toMatchSnapshot();
          expect(valuesArray).toEqual([userId]);
          return Promise.resolve(noMatchUserResponse);
        });
        global.pool.query.mockImplementationOnce((dbString, valuesArray) => {
          expect(dbString).toMatchSnapshot();
          expect(valuesArray).toEqual([userId]);
          return Promise.resolve(matchJoinResponse);
        });
        global.pool.query.mockImplementation((dbString, valuesArray) => {
          expect(dbString).toMatchSnapshot();
          expect(valuesArray).toEqual([userId, body.current]);
          return Promise.resolve(null);
        });

        const result = await updateSecurities(userId, body);

        expect(result).toEqual(null);
        expect(global.pool.query).toHaveBeenCalledTimes(3);
      });

      it('updates list, when join table response is empty', async () => {
        global.pool.query.mockImplementationOnce((dbString, valuesArray) => {
          expect(dbString).toMatchSnapshot();
          expect(valuesArray).toEqual([userId]);
          return Promise.resolve(matchUserResponse);
        });
        global.pool.query.mockImplementationOnce((dbString, valuesArray) => {
          expect(dbString).toMatchSnapshot();
          expect(valuesArray).toEqual([userId]);
          return Promise.resolve(emptyResponse);
        });
        global.pool.query.mockImplementationOnce((dbString, valuesArray) => {
          expect(dbString).toMatchSnapshot();
          expect(valuesArray).toEqual([userId]);
          return Promise.resolve(null);
        });
        global.pool.query.mockImplementation((dbString, valuesArray) => {
          expect(dbString).toMatchSnapshot();
          expect(valuesArray).toEqual([userId, body.list, indexArray]);
          return Promise.resolve(null);
        });

        const result = await updateSecurities(userId, body);

        expect(result).toEqual(null);
        expect(global.pool.query).toHaveBeenCalledTimes(4);
      });

      it('updates list, when joinList does not match', async () => {
        global.pool.query.mockImplementationOnce((dbString, valuesArray) => {
          expect(dbString).toMatchSnapshot();
          expect(valuesArray).toEqual([userId]);
          return Promise.resolve(matchUserResponse);
        });
        global.pool.query.mockImplementationOnce((dbString, valuesArray) => {
          expect(dbString).toMatchSnapshot();
          expect(valuesArray).toEqual([userId]);
          return Promise.resolve(noMatchJoinResponse);
        });
        global.pool.query.mockImplementationOnce((dbString, valuesArray) => {
          expect(dbString).toMatchSnapshot();
          expect(valuesArray).toEqual([userId]);
          return Promise.resolve(null);
        });
        global.pool.query.mockImplementation((dbString, valuesArray) => {
          expect(dbString).toMatchSnapshot();
          expect(valuesArray).toEqual([userId, body.list, indexArray]);
          return Promise.resolve(null);
        });

        const result = await updateSecurities(userId, body);

        expect(result).toEqual(null);
        expect(global.pool.query).toHaveBeenCalledTimes(4);
      });

      it('updates current security and list, when security_current and joinList do not match in each case', async () => {
        global.pool.query.mockImplementationOnce((dbString, valuesArray) => {
          expect(dbString).toMatchSnapshot();
          expect(valuesArray).toEqual([userId]);
          return Promise.resolve(noMatchUserResponse);
        });
        global.pool.query.mockImplementationOnce((dbString, valuesArray) => {
          expect(dbString).toMatchSnapshot();
          expect(valuesArray).toEqual([userId]);
          return Promise.resolve(noMatchJoinResponse);
        });
        global.pool.query.mockImplementationOnce((dbString, valuesArray) => {
          expect(dbString).toMatchSnapshot();
          expect(valuesArray).toEqual([userId, body.current]);
          return Promise.resolve(null);
        });
        global.pool.query.mockImplementationOnce((dbString, valuesArray) => {
          expect(dbString).toMatchSnapshot();
          expect(valuesArray).toEqual([userId]);
          return Promise.resolve(null);
        });
        global.pool.query.mockImplementation((dbString, valuesArray) => {
          expect(dbString).toMatchSnapshot();
          expect(valuesArray).toEqual([userId, body.list, indexArray]);
          return Promise.resolve(null);
        });

        const result = await updateSecurities(userId, body);

        expect(result).toEqual(null);
        expect(global.pool.query).toHaveBeenCalledTimes(5);
      });
    });
  });
});
