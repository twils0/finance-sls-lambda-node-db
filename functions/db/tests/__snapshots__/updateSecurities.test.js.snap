// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`functions db updateSecurities fails and returns unexpected error, when global.pool.query throws error fifth query 1`] = `
"SELECT security_current FROM \\"user\\".\\"user\\"
          WHERE user_id = $1::uuid;"
`;

exports[`functions db updateSecurities fails and returns unexpected error, when global.pool.query throws error fifth query 2`] = `
"SELECT security_id FROM \\"user\\".\\"join_user_security\\"
          WHERE user_id = $1::uuid;"
`;

exports[`functions db updateSecurities fails and returns unexpected error, when global.pool.query throws error fifth query 3`] = `
"UPDATE \\"user\\".\\"user\\"
          SET security_current = $2::uuid
          WHERE user_id = $1::uuid;"
`;

exports[`functions db updateSecurities fails and returns unexpected error, when global.pool.query throws error fifth query 4`] = `
"DELETE FROM \\"user\\".\\"join_user_security\\"
          WHERE user_id = $1::uuid;"
`;

exports[`functions db updateSecurities fails and returns unexpected error, when global.pool.query throws error fifth query 5`] = `
"INSERT INTO \\"user\\".\\"join_user_security\\" (user_id, security_id, security_order)
          VALUES($1::uuid, UNNEST($2::uuid[]), UNNEST($3::int2[]));"
`;

exports[`functions db updateSecurities fails and returns unexpected error, when global.pool.query throws error first query 1`] = `
"SELECT security_current FROM \\"user\\".\\"user\\"
          WHERE user_id = $1::uuid;"
`;

exports[`functions db updateSecurities fails and returns unexpected error, when global.pool.query throws error fourth query 1`] = `
"SELECT security_current FROM \\"user\\".\\"user\\"
          WHERE user_id = $1::uuid;"
`;

exports[`functions db updateSecurities fails and returns unexpected error, when global.pool.query throws error fourth query 2`] = `
"SELECT security_id FROM \\"user\\".\\"join_user_security\\"
          WHERE user_id = $1::uuid;"
`;

exports[`functions db updateSecurities fails and returns unexpected error, when global.pool.query throws error fourth query 3`] = `
"UPDATE \\"user\\".\\"user\\"
          SET security_current = $2::uuid
          WHERE user_id = $1::uuid;"
`;

exports[`functions db updateSecurities fails and returns unexpected error, when global.pool.query throws error fourth query 4`] = `
"DELETE FROM \\"user\\".\\"join_user_security\\"
          WHERE user_id = $1::uuid;"
`;

exports[`functions db updateSecurities fails and returns unexpected error, when global.pool.query throws error second query 1`] = `
"SELECT security_current FROM \\"user\\".\\"user\\"
          WHERE user_id = $1::uuid;"
`;

exports[`functions db updateSecurities fails and returns unexpected error, when global.pool.query throws error second query 2`] = `
"SELECT security_id FROM \\"user\\".\\"join_user_security\\"
          WHERE user_id = $1::uuid;"
`;

exports[`functions db updateSecurities fails and returns unexpected error, when global.pool.query throws error third query 1`] = `
"SELECT security_current FROM \\"user\\".\\"user\\"
          WHERE user_id = $1::uuid;"
`;

exports[`functions db updateSecurities fails and returns unexpected error, when global.pool.query throws error third query 2`] = `
"SELECT security_id FROM \\"user\\".\\"join_user_security\\"
          WHERE user_id = $1::uuid;"
`;

exports[`functions db updateSecurities fails and returns unexpected error, when global.pool.query throws error third query 3`] = `
"UPDATE \\"user\\".\\"user\\"
          SET security_current = $2::uuid
          WHERE user_id = $1::uuid;"
`;

exports[`functions db updateSecurities updates current security and list, when security_current and joinList do not match in each case 1`] = `
"SELECT security_current FROM \\"user\\".\\"user\\"
          WHERE user_id = $1::uuid;"
`;

exports[`functions db updateSecurities updates current security and list, when security_current and joinList do not match in each case 2`] = `
"SELECT security_id FROM \\"user\\".\\"join_user_security\\"
          WHERE user_id = $1::uuid;"
`;

exports[`functions db updateSecurities updates current security and list, when security_current and joinList do not match in each case 3`] = `
"UPDATE \\"user\\".\\"user\\"
          SET security_current = $2::uuid
          WHERE user_id = $1::uuid;"
`;

exports[`functions db updateSecurities updates current security and list, when security_current and joinList do not match in each case 4`] = `
"DELETE FROM \\"user\\".\\"join_user_security\\"
          WHERE user_id = $1::uuid;"
`;

exports[`functions db updateSecurities updates current security and list, when security_current and joinList do not match in each case 5`] = `
"INSERT INTO \\"user\\".\\"join_user_security\\" (user_id, security_id, security_order)
          VALUES($1::uuid, UNNEST($2::uuid[]), UNNEST($3::int2[]));"
`;

exports[`functions db updateSecurities updates current security, when securityCurrent does not match 1`] = `
"SELECT security_current FROM \\"user\\".\\"user\\"
          WHERE user_id = $1::uuid;"
`;

exports[`functions db updateSecurities updates current security, when securityCurrent does not match 2`] = `
"SELECT security_id FROM \\"user\\".\\"join_user_security\\"
          WHERE user_id = $1::uuid;"
`;

exports[`functions db updateSecurities updates current security, when securityCurrent does not match 3`] = `
"UPDATE \\"user\\".\\"user\\"
          SET security_current = $2::uuid
          WHERE user_id = $1::uuid;"
`;

exports[`functions db updateSecurities updates current security, when user table response is empty 1`] = `
"SELECT security_current FROM \\"user\\".\\"user\\"
          WHERE user_id = $1::uuid;"
`;

exports[`functions db updateSecurities updates current security, when user table response is empty 2`] = `
"SELECT security_id FROM \\"user\\".\\"join_user_security\\"
          WHERE user_id = $1::uuid;"
`;

exports[`functions db updateSecurities updates current security, when user table response is empty 3`] = `
"UPDATE \\"user\\".\\"user\\"
          SET security_current = $2::uuid
          WHERE user_id = $1::uuid;"
`;

exports[`functions db updateSecurities updates list, when join table response is empty 1`] = `
"SELECT security_current FROM \\"user\\".\\"user\\"
          WHERE user_id = $1::uuid;"
`;

exports[`functions db updateSecurities updates list, when join table response is empty 2`] = `
"SELECT security_id FROM \\"user\\".\\"join_user_security\\"
          WHERE user_id = $1::uuid;"
`;

exports[`functions db updateSecurities updates list, when join table response is empty 3`] = `
"DELETE FROM \\"user\\".\\"join_user_security\\"
          WHERE user_id = $1::uuid;"
`;

exports[`functions db updateSecurities updates list, when join table response is empty 4`] = `
"INSERT INTO \\"user\\".\\"join_user_security\\" (user_id, security_id, security_order)
          VALUES($1::uuid, UNNEST($2::uuid[]), UNNEST($3::int2[]));"
`;

exports[`functions db updateSecurities updates list, when joinList does not match 1`] = `
"SELECT security_current FROM \\"user\\".\\"user\\"
          WHERE user_id = $1::uuid;"
`;

exports[`functions db updateSecurities updates list, when joinList does not match 2`] = `
"SELECT security_id FROM \\"user\\".\\"join_user_security\\"
          WHERE user_id = $1::uuid;"
`;

exports[`functions db updateSecurities updates list, when joinList does not match 3`] = `
"DELETE FROM \\"user\\".\\"join_user_security\\"
          WHERE user_id = $1::uuid;"
`;

exports[`functions db updateSecurities updates list, when joinList does not match 4`] = `
"INSERT INTO \\"user\\".\\"join_user_security\\" (user_id, security_id, security_order)
          VALUES($1::uuid, UNNEST($2::uuid[]), UNNEST($3::int2[]));"
`;
